src/sbus/connection/sbus_watch.c

```diff
@@ -280,8 +290,27 @@ sbus_watch_remove(DBusWatch *dbus_watch, void *data)
 
     if (watch_fd->dbus_watch.read == NULL
             && watch_fd->dbus_watch.write == NULL) {
-        talloc_free(watch_fd->fdevent);
-        talloc_free(watch_fd);
+        /* libdbus doesn't need this watch{fd} anymore, so associated
+         * tevent_fd should be removed from monitoring at the spot.
+         *
+         * This first attempt to free will actually fail because tevent_fd is
+         * marked as busy in tevent_common_invoke_fd_handler() but it will
+         * stop fd monitoring in tevent loop.
+         * Memory will be eventually freed alongside with deallocation
+         * of parent context (watch_fd).
+         */
+        TALLOC_FREE(watch_fd->fdevent);
+        /* watch_fd itself can't be freed yet as it still may be referenced
+         * in the current context (for example in sbus_watch_handler())
+         * so instead schedule immediate event to delete it.
+         */
+        im = tevent_create_immediate(watch_fd->sbus_watch);
+        if (im == NULL) {
+            DEBUG(SSSDBG_FATAL_FAILURE, "Out of Memory!\n");
```

Comment:
I understand it is quite a corner case that really should not happen. But if it does, we will never free the watch. Can you extend `struct sbus_watch_fd` to hold the immediate event and create the event when creating the structure? I mean:

```diff
struct sbus_watch_fd {
    struct sbus_watch *sbus_watch;

    struct {
        DBusWatch *read;
        DBusWatch *write;
    } dbus_watch;

    int fd;
    struct tevent_fd *fdevent;
+  struct tevent_immediate *im;

    struct sbus_watch_fd *prev;
    struct sbus_watch_fd *next;
};

static struct sbus_watch_fd *
sbus_watch_get_by_fd(TALLOC_CTX *mem_ctx,
                     struct sbus_watch *watch,
                     int fd)
{
    struct sbus_watch_fd *watch_fd;

    /**
     * D-Bus may ask us to add a watch to a file descriptor that already had
     * a watch associated. If this is the case we return the existing context.
     */
    DLIST_FOR_EACH(watch_fd, watch->watch_list) {
        if (watch_fd->fd == fd) {
            return watch_fd;
        }
    }

    /* Create new one otherwise. */
    watch_fd = talloc_zero(mem_ctx, struct sbus_watch_fd);
    if (watch_fd == NULL) {
        DEBUG(SSSDBG_CRIT_FAILURE, "Out of memory!\n");
        return NULL;
    }
+ im = tevent_create_immediate(watch_fd);
+ if (im == NULL) {
+    ...
+  }

    talloc_set_destructor(watch_fd, sbus_watch_fd_destructor);

    watch_fd->sbus_watch = watch;
    watch_fd->fd = fd;

    return watch_fd;
}
```

Reply:
Honestly, I don't think this has real practical meaning. But such a change won't hurt either, so ok.

Reply:
> + im = tevent_create_immediate(watch_fd);

Is attaching `im` to the `watch_fd` mem-context safe?

`tevent_common_invoke_immediate_handler()` -> `free_sbus_watch()` -> `talloc_free(watch_fd)` -> `tevent_common_immediate_destructor(im)` fails, but `watch_fd` is freed.

Is `im` properly detached from parent context in this case (so that `tevent_common_invoke_immediate_handler()` cleaning up destroyed `im` with `talloc_free()` is fine)?

Reply:
I left attaching `im` to the `sbus_watch` context and explicit deletion in `free_sbus_watch()`.

If you think it is safe to attach `im` to `watch_fd` (and avoid need to free `im` explicitly) - please, let me know, I will amend patch again.